### ISSUE TITLE
Difference in init() alpha between src() and layer()?

### ISSUE THREAD

**Issue by ffd8** [hydra-synth/hydra-synth#153]
Was messing with some p5 + hydra and realized that when using p5 canvas with `src()` it doesn't have the expected alpha of the p5 canvas.. however if using it via `layer()`, it does?? This might be connected to #109 or it's something else? I wonder if the `init()` function is lacking some form of alpha or transparency?

Here's an example:
```js
p5 = new P5() 
p5.hide()

p5.draw = () => {
  p5.clear()
  p5.fill(255,0, 255, 50)
  p5.noStroke()
  p5.circle(p5.width/2, p5.height/2, 400)
}

s0.init({src: p5.canvas})

osc().layer(src(s0)).out() // p5 has alpha

//src(s0).out() // toggle on and alpha is missing??
```

expected or issue?

**Comment by geikha** [comment]
This problem unveils a very complex issue to be handled in future versions of the Hydra editor, since "fixing" would change the looks of older sketches, and therefore, a versioning system needs to be put in place.

This is discussed in -extensive- detail here:
https://github.com/hydra-synth/hydra-synth/issues/109

**Comment by ojack** [comment]
The source alpha is preserved identically in both cases, there is just a black background on the entire screen.  You can see this by changing the background color of the `body` element. 

```javascript
p5 = new P5() 
p5.hide()

p5.draw = () => {
  p5.clear()
  p5.fill(255,0, 255, 50)
  p5.noStroke()
  p5.circle(p5.width/2, p5.height/2, 400)
}

s0.init({src: p5.canvas})
src(s0).out()

document.body.style.backgroundColor = 'blue'
```

The discussion mentioned above by @geikha relates to compositing multiple layers together.

**Comment by geikha** [comment]
Both issues are mentioned in #109 !

This is the same as the part titled "The screen lies to you". Where some machines rendering a canvas won't premult while others (Mac) do, this is also why some sketches "don't work" on Mac, as they usually have something like `osc().sub(osc())` happening where alpha ends up being `1 - 1 = 0`. 

The same issue described here by @ffd8 appears when using a PNG with transparency, for the same reasons!

**Comment by ojack** [comment]
Oooh okay, i think I misunderstood!! I thought the behaviour mentioned in this issue was referring to the transparent areas around the shape, which for me on mac renders the same using both `src()` and layer. But you are talking about the semitransparent areas within the shape?

**Comment by geikha** [comment]
Windows 10 machine using an Intel integrated GPU:

As is:
![image](https://github.com/hydra-synth/hydra-synth/assets/56176668/9a6f31ba-e514-43f3-bd50-c1857ce97dbb)

Premultiplied:
![image](https://github.com/hydra-synth/hydra-synth/assets/56176668/12f1057d-33f1-49c2-8a79-54a4500675c2)

[open in hydra](https://hydra.ojack.xyz/?code=YXdhaXQlMjBsb2FkU2NyaXB0KCUyMmh0dHBzJTNBJTJGJTJGaHlwZXItaHlkcmEuZ2xpdGNoLm1lJTJGaHlkcmEtZ2xzbC5qcyUyMiklMEElMEFwNSUyMCUzRCUyMG5ldyUyMFA1KCklMjAlMEFwNS5oaWRlKCklMEElMEFwNS5kcmF3JTIwJTNEJTIwKCklMjAlM0QlM0UlMjAlN0IlMEElMjAlMjBwNS5jbGVhcigpJTBBJTIwJTIwcDUuZmlsbCgyNTUlMkMwJTJDJTIwMjU1JTJDJTIwNTApJTBBJTIwJTIwcDUubm9TdHJva2UoKSUwQSUyMCUyMHA1LmNpcmNsZShwNS53aWR0aCUyRjIlMkMlMjBwNS5oZWlnaHQlMkYyJTJDJTIwNDAwKSUwQSU3RCUwQSUwQXMwLmluaXQoJTdCc3JjJTNBJTIwcDUuY2FudmFzJTdEKSUwQSUwQW9zYygpLmxheWVyKHNyYyhzMCkpLm91dCgpJTIwJTJGJTJGJTIwcDUlMjBoYXMlMjBhbHBoYSUwQSUwQXNyYyhzMCklMEElMjAlMjAlMDklMkYlMkYuZ2xzbENvbG9yKCd2ZWM0KGMwLnJnYipjMC5hJTJDJTIwYzAuYSknKSUyMCUyRiUyRiUyMHByZW11bHRpcGx5JTIwYWxwaGElMjBiZWZvcmUlMjBvdXRwdXR0aW5nJTBBJTIwJTIwJTA5Lm91dCgpJTIwJTJGJTJGJTIwdG9nZ2xlJTIwb24lMjBhbmQlMjBhbHBoYSUyMGlzJTIwbWlzc2luZyUzRiUzRg%3D%3D)

**Comment by ffd8** [comment]
Exactly, it's about a png or transparent p5 source not displaying with that transparency when using `src()`. Indeed just tested the example just above and that `glslColor()` solves the issue. I guess some love the fact it's opaque with `src()` versus `layer()` and that's where it needs to be versioned? or is that about something else? I see the issue more obviously when drawing random circles across the page (what I was initially doing when seeing the issue) – it goes from the alpha overlap of shapes with `layer()` to solid quickly full display with `src()`:

```
p5 = new P5() 
p5.hide()

p5.draw = () => {
  //p5.clear()
  p5.fill(255, 50)
  p5.noStroke()
  p5.circle(p5.random(p5.width), p5.random(p5.height), 100)
}

s0.init({src: p5.canvas})

osc().layer(src(s0)).out() // p5 has alpha

// src(s0).out() // toggle on to see diff

document.body.style.backgroundColor = 'blue'
```

Would be interesting to know in what contexts it's preferred to have that 'opaque' quality of `src()`?

**Comment by ojack** [comment]
> Would be interesting to know in what contexts it's preferred to have that 'opaque' quality of src()?

I don't think it is preferable in any case tbh! just that we haven't found away to fix it without changing the behavior of many other things.  For now, I recommend using the workarounds shared by @geikha here and in the other issue. 

> Interestingly enough, when i 'CTRL + SHIFT + S' to capture the image of what i see.. the display of it being 'opaque' with just src(s0).out() – still saves with the transparent qualities??

That is true for me too!!! so weird!!!

**Comment by ojack** [comment]
this article seems directly related to this issue: https://webglfundamentals.org/webgl/lessons/webgl-text-texture.html

it recomends using 
```javascript
gl.pixelStorei(gl.UNPACK_PREMULTIPLY_ALPHA_WEBGL, true);
```
when binding the texture 

and 
```javascript
gl.blendFunc(gl.ONE, gl.ONE_MINUS_SRC_ALPHA);
```
overall....but I can't say I understand everything happening here.

**Comment by ojack** [comment]
and also this one which has some other webgl flags: https://webglfundamentals.org/webgl/lessons/webgl-and-alpha.html

**Comment by geikha** [comment]
> Interestingly enough, when i 'CTRL + SHIFT + S' to capture the image of what i see.. the display of it being 'opaque' with just src(s0).out() – still saves with the transparent qualities??
@ffd8 

Because this is basically a rendering error. The image is still transparent, you're just seeing on screen the effect of WebGL trying to render a premultiplied transparent image but getting a straight alpha image. Hydra should either have all its compositing functions expect premultiplied sources (this includes telling WebGL to premultiply external sources that use straight alpha such as Canvas and PNG images), or use straight alpha all around and premultiply at the end, just before rendering to screen. 

As mentioned above, changing this behaviour would "break" older sketches, specially considering how minor changes can affect chaotic feedback systems which are very common in Hydra. Therefore we just can't fix this until there's a versioning system for the Hydra editor which can read the timestamp of each saved sketch and use the corresponding version.

Btw, if we continue discussing this issue here it'll just turn into a copy of #109 😅
I would suggest going there if there are any other thoughts about alpha compositing

**Comment by ffd8** [comment]
Rightio – many thanks @geikha + @ojack for taking time to look into this! The whole process flys waay over my head and a shame causes such problems for previous sketches (rereading #109 a few times). Maybe an easy 'going forward' thing – would be to pass a flag along with `src(tex, alphaToggle)` (and similar functions where this applies) - where that 2nd value is a 0 or 1 or something else, to find its way down the pipeline to use `premultiplied alpha` or `straight alpha`? Then it would be backwards compatible.. previous sketches would default either '1' or '0' (haha not sure which should represent which) – and newer sketches could toggle it as needed?
